srcs/serial_control.cpp

```diff
@@ -0,0 +1,151 @@
+/******************************************************************************
+ * @author Makers For Life
+ * @copyright Copyright (c) 2020 Makers For Life
+ * @file serial_control.cpp
+ * @brief Handle control protocol on the serial input
+ *****************************************************************************/
+
+#pragma once
+
+#include "../includes/config.h"
+
+// INCLUDES ===================================================================
+
+// Associated header
+#include "../includes/serial_control.h"
+
+// Externals
+#if HARDWARE_VERSION == 2 || HARDWARE_VERSION == 3
+#include "Arduino.h"
+#include "CRC32.h"
+#endif
+
+/// Internals
+#include "../includes/pressure_controller.h"
+
+// GLOBAL ITEMS ==============================================================
+
+#define CONTROL_HEADER_SIZE 2
+static const uint8_t header[CONTROL_HEADER_SIZE] = {0x05, 0x0A};
+#define CONTROL_FOOTER_SIZE 2
+static const uint8_t footer[CONTROL_FOOTER_SIZE] = {0x50, 0xA0};
+
+// FUNCTIONS ==================================================================
+
+/**
+ * Convert an array of 2 bytes to a u16
+ *
+ * @param bytes Array of 2 elements
+ * @return The corresponding u16 number
+ */
+// cppcheck-suppress unusedFunction
+uint16_t toU16(byte bytes[]) {
+    uint16_t num = (bytes[0] << 8) + bytes[1];
+    return num;
+}
+
+/**
+ * Convert an array of 4 bytes to a u32
+ *
+ * @param bytes Array of 4 elements
+ * @return The corresponding u32 number
+ */
+// cppcheck-suppress unusedFunction
+uint32_t toU32(byte bytes[]) {
+    uint32_t num = (bytes[0] << 24) + (bytes[1] << 16) + (bytes[2] << 8) + bytes[3];
+    return num;
+}
+
+// cppcheck-suppress unusedFunction
+void serialControlLoop() {
+#if HARDWARE_VERSION == 2 || HARDWARE_VERSION == 3
+    // Let's note this current time to avoid blocking too long here
+    int time = millis();
+
+    // We need to ensure we received the whole message
+    while (((time + 2) >= millis()) && (Serial6.available() >= 11)) {
```

Comment:
humm, j'aime pas cette boucle, surtout avec des continue à l'intérieur. comment on peut s'assurer de sortir du while en cas de spam de l'uart ?

Cette boucle peut générer une dérive temporelle non bornée et pourquoi time+2 >millis ? si la boucle est indispensable et que la condition temporelle l'est aussi, peut-on l'écrire de manière à l'invalider à un moment ?

Reply:
J'ai mis une boucle ici pour être capable de gérer le fait qu'il pourrait y avoir autre chose que des messages qui soient reçus, par exemple des octets isolés. Par ailleurs, je voulais pouvoir prendre en compte rapidement plusieurs messages qui seraient envoyés à la suite. L'objectif c'est que ça soit le plus réactif possible (notamment à renvoyer un message ACK à la UI) sans impacter la respiration.

- on arrête d'essayer de traiter des messages si jamais :
    - on a moins de 11 octets en buffer (= on est sûr qu'on n'a aucune chance d'avoir le message en entier donc on verra plus tard si on a reçu plus d'octets)
    - ou si on a déjà passé plus de 2 ms à essayer (valeur arbitraire, pour éviter de prendre de la dérive temporelle sur les fonctions de respiration)
- les `continue` permettent de réessayer de traiter des messages si on a eu une erreur (mauvais CRC, mauvaise structure, pas les bons entêtes) pour être plus réactif que si on attend juste le tick d'après ; ce "retry" est limité dans le temps par les 2 critères évoqués ci-dessus.

Mon objectif ici est de tester ça sur un proto pour s'assurer que la respiration continue à fonctionner même dans le cas d'un RPi malveillant. J'ai déjà écrit du code pour essayer de flooder le MCU :  [makers-for-life/makair-telemetry#12](https://github.com/makers-for-life/makair-telemetry/pull/12)

Tu penses que cette approche est pétée ?